### PR TITLE
[Workflow] Add colors to workflow dumps

### LINKED ISSUE #28874: [Workflow] Add colors to workflow dumps
**Description**  
The PR _[Workflow] Add a MetadataStore to fetch some metadata #26092_ has been merged so we can add some metadata to our workflows.

The text of the PR shows an example with `color: grey`, so it looks like the support of color was planned.

When I [dump the workflow](https://symfony.com/doc/current/workflow/dumping-workflows.html) with both formats, there are no color on the images, like the examples in the documentation.

How can we add colors to the dumped workflow? Is the feature implemented yet?

See also https://github.com/symfony/symfony-docs/pull/9476 if it needs it be documented.

**Example**  

I use `symfony/workflow:4.1.3` (the problem is the same with `4.1.6`).

I have a workflow like this:

```yaml
framework:
    workflows:
        my_entity:
            type: 'state_machine'
            marking_store:
                type: 'single_state'
                arguments:
                    - 'state'
            supports:
                - App\Entity\MyEntity
            initial_place: created
            places:
                created:
                    metadata:
                        description: Entity created, waiting for next step
                        color: green
            # …
```

And the output doesn't use the colors or the description.

### PR DESCRIPTION
**Pull request by alexisLefebvre**
Fixes #28874

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28874, replaces #28933
| License       | MIT
| Doc PR        | TODO, requires https://github.com/symfony/symfony-docs/pull/9476

Fetch data with the `MetadataStore` from #26092 in order to add colors to the dumps.

Example of configuration:

```yaml
            transitions:
                submit:
                    from: start
                    to: travis
                    metadata:
                        title: transition submit title
                        dump_style:
                            label: 'My custom label'
                            arrow_color: '#0088FF'
                            label_color: 'Red'
```

This code was developed as a bundle, examples can be found on its repository: https://github.com/alexislefebvre/SymfonyWorkflowStyleBundle

### CODE REVIEW

File: src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php [hunk 1]
```diff
@@ -82,19 +90,35 @@ public function dump(Definition $definition, Marking $marking = null, array $opt
...
+
+                    $transitionColor = '';
+
+                    if (isset($style['arrow_color'])) {
+                        $transitionColor = $this->getTransitionColor($style['arrow_color']);
+                    }
+
                     if ($this->isWorkflowTransitionType()) {
+                        $transitionLabel = '';
+                        // Add label only it has a style
```
Comment (javiereguiluz) [inline src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php]: `... only it ...` -> `... only if it ...`

File: src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php [hunk 2]
```diff
@@ -160,4 +198,78 @@ private function escape(string $string): string
...
+            $color = '#'.$color;
+        }
+
+        return sprintf('[%s]', $color);
+    }
+
+    /**
+     * Remove “#“ from start of the color name so it can be used as an identifier.
+     */
+    private function getColorId(string $color): string
```
Comment (javiereguiluz) [inline src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php]: Can we simplify this method by something like this?

```php
return ltrim($color, '#');
```

File: src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php [hunk 3]
```diff
@@ -160,4 +198,78 @@ private function escape(string $string): string
...
+        return $this->workflowMetadata->getMetadata('dump_style', $place) ?? array();
+    }
+
+    private function getTransitionStyle(Transition $transition): array
+    {
+        return $this->workflowMetadata->getMetadata('dump_style', $transition) ?? array();
+    }
+
+    /**
+     * Color in transition must be prefixed with “#”.
```
Comment (javiereguiluz) [inline src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php]: I think this comment doesn't describe the method but the inner code, so I'd move the comment inside the method. Same for `getColorId()`'s comment.

File: src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php [hunk 4]
```diff
@@ -160,4 +198,78 @@ private function escape(string $string): string
...
+     */
+    private function getTransitionColor(string $color): string
```
Comment (javiereguiluz) [inline src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php]: I find this `#` trimming a bit confusing. In some test you use `Grey` and `Red` as colors, but here we say that colors must be in HEX format (they must have the leading `#` symbol). So, what color formats are supported? Thanks!

File: src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php [hunk 5]
```diff
@@ -26,12 +29,17 @@
...
+    use GetMetadataTrait;
+
     protected static $defaultOptions = array(
         'graph' => array('ratio' => 'compress', 'rankdir' => 'LR'),
         'node' => array('fontsize' => 9, 'fontname' => 'Arial', 'color' => '#333333', 'fillcolor' => 'lightblue', 'fixedsize' => 'false', 'width' => 1),
         'edge' => array('fontsize' => 9, 'fontname' => 'Arial', 'color' => '#333333', 'arrowhead' => 'normal', 'arrowsize' => 0.5),
     );
 
+    /** @var MetadataStoreInterface */
+    protected $workflowMetadata;
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: This class is a service. You should not store specific data in it.

**Review by lyrixx — COMMENTED** [review]
I did not tested it, but It's a very good start

File: src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php (hunk 5 above)
Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: Could you please suggest a better solution? I don't know how to fix this, using an interface maybe?

Comment (lyrixx) [reply src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: you just need to pass the workflow metadata to all method that need to use it

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/workflows.php [hunk 6]
```diff
@@ -14,12 +14,19 @@
...
-                'draft',
-                'wait_for_journalist',
-                'approved_by_journalist',
-                'wait_for_spellchecker',
-                'approved_by_spellchecker',
-                'published',
+                'draft' => array(),
+                'wait_for_journalist' => array(
+                    'metadata' => array(
+                        'dump_style' => array(
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/workflows.php]: the extra level is not really needed is it. We could reserve 1st level keywords (label, color, description) for any generic purpose.

File: src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php (hunk 4 above)
Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php]: Without `#`, PUML throws an error, so we have to add to use the following notation:

```
"b" -[#Pink]-> "t2"
```

**Review by Kronhyx — APPROVED** [review]

File: src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php (hunk 5 above)
Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: Thanks, the code was simpler after removing this. :+1:

**Comment by alexislefebvre** [comment]
Can someone please remove the `Deprecation` tag? I forgot to change the value when creating the PR.

File: src/Symfony/Component/Workflow/CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,11 @@ CHANGELOG
 -----
 
  * Trigger `entered` event for subject entering in the Workflow for the first time
+ * Added dump styling of states and transitions through `metadata`,
+   configuration of places is:
+   `label` and `description` (only for PUML) (text), `bg_color` (`#RRGGBB` or CSS color name)
```
Comment (noniagriconomie) [inline src/Symfony/Component/Workflow/CHANGELOG.md]: hello, i think it will be better to prefix by **dump_** those metadata, as they are specific for the dump feature
wdyt?

Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: This is a good question because it may conflict with an existing configs in some projects. We chose to use a simple syntax: https://github.com/symfony/symfony/pull/29538#discussion_r241770962

**Comment by nicolas-grekas** [comment]
rebase needed due to short arrays.

**Comment by alexislefebvre** [comment]
Thanks for hint, the PR has been rebased.

**Comment by fabpot** [comment]
@lyrixx I think we need your help to move forward here.

**Review by lyrixx — APPROVED** [review]
:+1:

**Comment by lyrixx** [comment]
Thanks @alexislefebvre . This is very Good. Could you rebase before the merge ? (I think this will conflict with #30468 but I will take care of it.

Could someone review English textes (PHP Doc, Changelog)? :pray: Thanks

**Comment by alexislefebvre** [comment]
Rebase done. :heavy_check_mark: 

Thanks for the follow-up.

File: src/Symfony/Component/Workflow/CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,11 @@ CHANGELOG
 
  * Trigger `entered` event for subject entering in the Workflow for the first time
  * Added a context to `Workflow::apply()`. The `MethodMarkingStore` could be used to leverage this feature.
+ * Added dump styling of states and transitions through `metadata`,
+   configuration of places is:
+   `label` and `description` (only for PUML) (text), `bg_color` (`#RRGGBB` or CSS color name)
+   and configuration of transitions is:
+   `label` (text), `color` and `arrow_color` (`#RRGGBB` or CSS color name)
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/CHANGELOG.md]: Instead of describing in plain English, I suggest to write a PHP example on how to use the new feature, it should be more effective.

Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: Ok, we can take examples from the test file: https://github.com/symfony/symfony/pull/29538/files?file-filters%5B%5D=.md&file-filters%5B%5D=.php#diff-f7d48f660a821958cbe877f5eb738de5

And this part part should be moved to the documentation.

Comment (lyrixx) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: let's do it!

Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: It's done.

File: src/Symfony/Component/Workflow/CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,26 @@ CHANGELOG
...
+            'arrow_color' => '#00ff00',
+        ];
+        $inMemoryMetadataStore = new InMemoryMetadataStore([], [], $transitionsMetadata);
+        ```
+
+   5. Pass metadata store as the fourth argument of `Definition`:
+
+        ```
+        return new Definition($places, $transitions, null, $inMemoryMetadataStore);
+        ```
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/CHANGELOG.md]: What about plain PHP (without anything that is not strictly needed as docs are where this should be explained in details):

```php
        use Symfony\Component\Workflow\Definition;
        use ...;

        $transitionsMetadata = new \SplObjectStorage();
        $transitionsMetadata[$transition] = [
            'color' => 'Red',
            'arrow_color' => '#00ff00',
        ];
        $inMemoryMetadataStore = new InMemoryMetadataStore([], [], $transitionsMetadata);

      $definition = new Definition($places, $transitions, null, $inMemoryMetadataStore);
```

Comment (alexislefebvre) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: Thanks for the advice, it's done.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexislefebvre.

File: src/Symfony/Component/Workflow/CHANGELOG.md (hunk 9 above)
Comment (fabpot) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: I've tweaked the README in ddf0c951e2
